### PR TITLE
[SandboxVec][BottomUpVec] Generate vector instructions

### PR DESCRIPTION
**Pull request by vporpo**
This patch implements some very basic code generation, for some opcodes.

### CODE REVIEW

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/VecUtils.h [hunk 1]
```diff
@@ -66,6 +66,40 @@ class VecUtils {
     }
     return true;
   }
+
+  /// \Returns the number of vector lanes of \p Ty or 1 if not a vector.
+  /// NOTE: It asserts that \p Ty is a fixed vector type.
+  static unsigned getNumLanes(Type *Ty) {
+    assert(!isa<ScalableVectorType>(Ty) && "Expect fixed vector");
```
Comment (slackito) [inline llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/VecUtils.h:73]: Nit: "Expected scalar or fixed vector type"?

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/VecUtils.h [hunk 2]
```diff
@@ -66,6 +66,40 @@ class VecUtils {
...
+    if (!isa<FixedVectorType>(Ty))
+      return 1;
+    return cast<FixedVectorType>(Ty)->getNumElements();
+  }
+
+  /// \Returns the expected vector lanes of \p V or 1 if not a vector.
+  /// NOTE: It asserts that \p V is a fixed vector.
+  static int getNumLanes(Value *V) {
```
Comment (slackito) [inline llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/VecUtils.h:81]: Nit: return `unsigned` to match the return type of `VecUtils::getNumLanes(Type*)`

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/VecUtils.h [hunk 3]
```diff
@@ -66,6 +66,40 @@ class VecUtils {
...
     return true;
   }
+
+  /// \Returns the number of vector lanes of \p Ty or 1 if not a vector.
+  /// NOTE: It asserts that \p Ty is a fixed vector type.
+  static unsigned getNumLanes(Type *Ty) {
+    assert(!isa<ScalableVectorType>(Ty) && "Expect fixed vector");
+    if (!isa<FixedVectorType>(Ty))
+      return 1;
+    return cast<FixedVectorType>(Ty)->getNumElements();
```
Comment (slackito) [inline llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/VecUtils.h:76]: This isa + cast could be a dyn_cast.

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Passes/BottomUpVec.h [hunk 4]
```diff
@@ -25,7 +25,9 @@ namespace llvm::sandboxir {
 class BottomUpVec final : public FunctionPass {
   bool Change = false;
   std::unique_ptr<LegalityAnalysis> Legality;
-  void vectorizeRec(ArrayRef<Value *> Bndl);
+
+  Value *createVectorInstr(ArrayRef<Value *> Bndl, ArrayRef<Value *> Operands);
```
Comment (slackito) [inline llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Passes/BottomUpVec.h:29]: A one-line comment explaining what the arguments mean would be nice.

**Review by slackito — APPROVED** [review]
Looks good to me, just a few nits.

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/VecUtils.h (hunk 3 above)
Comment (vporpo) [reply llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/VecUtils.h]: Done

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/VecUtils.h (hunk 2 above)
Comment (vporpo) [reply llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/VecUtils.h]: Done

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/VecUtils.h (hunk 1 above)
Comment (vporpo) [reply llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/VecUtils.h]: Done

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Passes/BottomUpVec.h (hunk 4 above)
Comment (vporpo) [reply llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/Passes/BottomUpVec.h:31]: Done

**Pull request merged**